rust/kernel/workqueue.rs

```diff
@@ -391,7 +391,7 @@ impl Work {
             // SAFETY: When the work was queued, a call to `into_raw` was made. We just canceled
             // the work without it having the chance to run, so we need to explicitly destroy this
             // reference (which would have happened in `work_func` if it did run).
-            unsafe { Ref::from_raw(&*self) };
+            unsafe { Ref::from_raw(self) };
```

Comment:
Pre-existing, but I think this is actually wrong. `Work` is not the thing put in a `Ref`, but the type containing `Work`. As such `self` should be value containing `Work`. Could you please take a look @wedsonaf?

Reply:
You're right, we need to do the same pointer arithmetic we do in `work_func` below to get to the right ptr.

Nice catch, @bjorn3 

Now we don't have the type here anymore, so we either need to make `Work` generic or just store the offset is some field.

Reply:
But that would be unsafe. Users could use the wrong adapter.

Reply:
oops, sorry, I see now that it's passing the outer object.